tools/shared.py

```diff
@@ -2564,7 +2564,8 @@ def metadce(js_file, wasm_file, minify_whitespace, debug_info):
     assert Settings.DECLARE_ASM_MODULE_EXPORTS, 'Internal error: Meta-DCE is currently not compatible with -s DECLARE_ASM_MODULE_EXPORTS=0, build should have occurred with -s DECLARE_ASM_MODULE_EXPORTS=1 if we reach here!'
     temp_files = configuration.get_temp_files()
     # first, get the JS part of the graph
-    txt = Building.js_optimizer_no_asmjs(js_file, ['emitDCEGraph', 'noEmitAst'], return_output=True)
+    extra_info = '{ "exports": [' + ','.join(map(lambda x: '["' + x + '","' + x + '"]', Settings.MODULE_EXPORTS)) + ']}'
```

Comment:
i don't understand why the module exports are special here?

Reply:
In DECLARE_ASM_MODULE_EXPORTS=0 builds, JS optimizer cannot traverse and do AST matching to get the asm.js module exports, because they are not declared, so we pass them there directly. This could be gated behind `DECLARE_ASM_MODULE_EXPORTS=0`, but it is just simpler to unconditionally do this to have fewer branches (one could argue even to delete the whole AST matching business of figuring out the exports, that needs to walk through the whole JS code, to get info that is readily available here - but I don't want to disrupt everything here, so made the two approaches complementary)